### PR TITLE
[ADD] payment_mollie: added mollie payment acquirer

### PR DESCRIPTION
**Pull request by pga-odoo**
Added mollie payment gateway. Mollie is a major payment provider in Europe and it is famous for providing multiple payments methods by a single acquirer. 

With mollie, you can use the following methods,
- Credit cards with inline component
- Apple Pay
- SOFORT banking
- SEPA Direct Debit
- SEPA Bank Transfer
- Paypal
- Klarna: Pay later
- Klarna: Slice it
- iDEAL
- Giropay
- EPS
- Bancontact
- KBC/CBC Payment Button
- Belfius Pay Button
- Eco-, gift- and meal vouchers
- Przelewy24
And many more

image
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pga-odoo** [comment]
Let me know if you have any queries :wink:

**Comment by Yenthe666** [comment]
Huh, this is interesting.. Last time I discussed & proposed it with Odoo (just over a year ago) they didn't want Mollie in and denied it. I'm curious what made them change their mind this time..

**Comment by fmdl** [comment]
there are made odoo module him self : https://github.com/mollie/mollie-odoo

it is possible to backport in V14 ?

**Comment by Yenthe666** [comment]
@fmdl it is actually @pga-odoo that works on this module lately :) & I've already proposed it for V14 (even before V14 was released) and Odoo SA refused it so 🤷

**Comment by AntoineVDV** [comment]
Hello, if this PR gets merged it will unfortunately not be backported to V14. This is mainly because the `payment` module was heavily refactored in 14.3 to make it easier to integrate new payment solutions. Integrating a payment provider in an earlier version is more complicated, is not always possible, and means supporting two very different versions of the same module.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Hello, @pga-odoo @kig-odoo first of all, kudos for writing clean and detailed code :clap:  It was really helpful for the review.

I had a deep look at this PR and came to the conclusion that a custom embedded payment method selection form is not the best way to integrate Mollie with Odoo. I believe a regular [Mollie Checkout](https://docs.mollie.com/payments/hosted-checkout) integration would offer almost the same payment options while requiring much fewer hacks and customizations. I understand that switching from one integration to another can be frustrating but it can also be exciting to achieve the same results with a fraction of the code.

From my point of view, the main problem with the proposed integration is that Mollie doesn't offer a pre-made widget for the selection of the payment method (PM) like Adyen or Stripe do. This makes it mandatory to go through a series of customizations that are generally best to avoid if we want to keep things simple and maintainable:
- synchronization of available PMs and their translations
- mixing of Mollie PMs and Mollie PM issuers in the `payment.icon` table
- mapping of Mollie PMs with Odoo's payment icons
- hack of the payment form to build a PM selection form and present PMs as acquirers
- implementation of issuers selection

This is *a lot* considering that pretty much all payment methods require the customer to be redirected to Mollie to finalize the payment.

I concede that switching to the Mollie Checkout integration also prevents from integrating more advanced features such as the configuration of vouchers or the integration with Sales, but we'd rather not have these in standard Odoo for now.

Note: The deadline for merging this PR has been pushed back one month to the 2nd of September.

**Comment by pga-odoo** [comment]
@AntoineVDV Thank you for the feedback. We will start working on it and update the code based on **Mollie Checkout integration** and we will remove all extra things. 

Are we still want to use Order API and Payment API both or just Payment API? Order API is used in sales orders and enables some extra methods like Klarna Pay later 

Just putting this information here as an alternative, in case you like it, I can do something like this too. There is an alternative possible by adding a dropdown like the given image. And it depends on [Methods API](https://docs.mollie.com/reference/v2/methods-api/list-methods) to check available methods, so we can remove several things like the method table, issuer table, some JS code, PM selection form changes, and icons.  image

**Comment by pga-odoo** [comment]
@AntoineVDV deadline push backed to 1 month so is the deadline is 2nd September or 2nd August?

**Comment by AntoineVDV** [comment]
@pga-odoo Let's stick to the Payment API only for now. We see the Orders API as nice to have but a bit overkill. Moreover, integrating with Sales would require an additional module `sale_payment_mollie` as `payment_mollie` doesn't have any dependence on `sale`.

2nd of September*

**Comment by pga-odoo** [comment]
@AntoineVDV thanks for the update.

I will let you know once it is completed according to your feedback.

**Comment by pga-odoo** [comment]
@AntoineVDV I have updated the code as per your feedback. Now module is working based on **Mollie Checkout** 

I have also removed all extra code related to checkout customization and order API etc.  Compared to the previous implementation, code is reduced by 70% now :smile: 

Let me know if you want me to change anything else.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Hello, please find my latest review comments below :)

Also, you should probably add an override of `_get_compatible_acquirers` to unlist Mollie acquirers from the payment form if the currency is not supported by Mollie. You can check how it is done in `payment_sips`.

File: addons/payment_mollie/__manifest__.py
Comment (AntoineVDV) [inline addons/payment_mollie/__manifest__.py:12]:
```suggestion
    'website': 'https://www.mollie.com',
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:123]:
```suggestion
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [reply addons/payment_mollie/models/payment_transaction.py:101]: Alright, no need to this the form then if it has no input and if the action is not known in advance.

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:105]:
```suggestion
    def _mollie_prepare_payment_request_payload(self):
        """ Create the payload for the payment request based on the transaction values.

        :return: The request payload
        :rtype: dict
        """
```
No need to specify that it is split for the tests. There is nothing wrong with extracting this method.

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:98]: Please move this method after `_get_specific_rendering_values`.

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:109]:
```suggestion
        user_lang = self.env.context.get('lang')
        base_url = self.acquirer_id.get_base_url()
        redirect_url = urls.url_join(base_url, MollieController._return_url)
        webhook_url = urls.url_join(base_url, MollieController._notify_url)
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:111]:
```suggestion
        return {
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:35]:
```suggestion
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:36]:
```suggestion
        payload = self._mollie_prepare_payment_payload()
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [reply addons/payment_mollie/models/payment_transaction.py]: This is ok. Tests should always include a webhook anyway.

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [reply addons/payment_mollie/models/payment_transaction.py]: Alright, I didn't see that it was used for the request.

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:81]:
```suggestion
        payment_data = self.acquirer_id._mollie_make_request(
            f'/payments/{self.acquirer_reference}', method="GET"
        )
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:111]:
```suggestion
        return {
```

File: addons/payment_mollie/models/payment_transaction.py
Comment (AntoineVDV) [inline addons/payment_mollie/models/payment_transaction.py:37]:
```suggestion
        _logger.info("sending '/payments' request for link creation:\n%s", pprint.pformat(payload))
        payment_data = self.acquirer_id._mollie_make_request('/payments', data=mollie_payload)
```

**Comment by Feyensv** [comment]
Note that the `ci/style` check should also be green on your PR before the merge (`ci/security` check is overridden manually after review from our internal security team).
It's red because of 5 unused imports remaining in your code ;) (Details [here](http://runbot.odoo.com/runbot/build/9673707))

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:9]:
```suggestion
from odoo.exceptions import ValidationError
from odoo.http import request, Response
```

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:17]:
```suggestion

    @http.route(
```

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:23]:
```suggestion

        The route is flagged with `save_session=False` to prevent Odoo from assigning a new session
```

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:30]:
```suggestion

        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
```

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:33]:
```suggestion
```

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:38]:
```suggestion
    @http.route(_notify_url, type='http', auth='public', methods=['POST'], csrf=False)
```

File: addons/payment_mollie/controllers/main.py
Comment (AntoineVDV) [inline addons/payment_mollie/controllers/main.py:41]:
```suggestion

        :param dict data: The feedback data (only `id`) and the transaction reference (`ref`)
```

File: addons/payment_mollie/views/payment_views.xml
Comment (AntoineVDV) [inline addons/payment_mollie/views/payment_views.xml:10]:
```suggestion
                    <field name="mollie_api_key" attrs="{'required': [('provider', '=', 'mollie'), ('state', '!=', 'disabled')]}" password="True"/>
```

File: addons/payment_mollie/tests/test_mollie.py
Comment (AntoineVDV) [inline addons/payment_mollie/tests/test_mollie.py:14]:
```suggestion
    def test_payment_request_payload_values(self):
```

**Review by AntoineVDV — CHANGES_REQUESTED** [review]
Please have a look at the "CI/style" check at the bottom of the PR. There are a few imports to clean for the check to pass.

File: addons/payment_mollie/const.py
Comment (pga-odoo) [reply addons/payment_mollie/const.py:3]: it is  ISO 15897 locale. I have added mollies link which have locale details

File: addons/payment_mollie/models/payment_acquirer.py
Comment (pga-odoo) [reply addons/payment_mollie/models/payment_acquirer.py]: I use `API key` because there is one field in the `payment_adyen` module which is also having the same label and it's giving this warning 

> Two fields (adyen_api_key, mollie_api_key) of payment.acquirer() have the same label: API Key

**Comment by AntoineVDV** [comment]
@pga-odoo Please reset my request for review when you're ready.

**Comment by pga-odoo** [comment]
@AntoineVDV I have made changes according to your feedback.  I have added comments for some feedbacks. 

Waiting for runbot to pass all tests :crossed_fingers: 

Let me know if you need anything from my side :slightly_smiling_face:

File: addons/payment_mollie/models/payment_acquirer.py
Comment (AntoineVDV) [reply addons/payment_mollie/models/payment_acquirer.py]: Indeed, then please change the `string` in the field definition to `Mollie API Key` and to `API Key` in the view.

File: addons/payment_mollie/models/payment_acquirer.py
Comment (Feyensv) [inline addons/payment_mollie/models/payment_acquirer.py:40]:
```suggestion
        if currency and currency.name not in SUPPORTED_CURRENCIES:
            acquirers = acquirers.filtered(lambda a: a.provider != 'mollie')
```
Move the generic condition outside the filtered lambda function, no need to evaluate it for each acquirer in the recordset ;).

File: addons/payment_mollie/models/payment_transaction.py
Comment (Feyensv) [inline addons/payment_mollie/models/payment_transaction.py:28]:
```suggestion
        :return: The dict of acquirer-specific rendering values
```

File: addons/payment_mollie/models/payment_acquirer.py
Comment (pga-odoo) [reply addons/payment_mollie/models/payment_acquirer.py]: Done :+1:

**Comment by pga-odoo** [comment]
@AntoineVDV @Feyensv tests are green now.
I have also updated everything according to your last feedback.

**Review by AntoineVDV — APPROVED** [review]
@pga-odoo Thank you for your quick work, it looks all good for me now :)

**Comment by AntoineVDV** [comment]
@odoo/rd-security could you please review this one?

In particular, I'm not sure if the `except requests.exceptions.RequestException` (see CI logs) could this be an issue? We usually only catch specific subclasses and let other exceptions bubble up.

**Comment by xmo-odoo** [comment]
> In particular, I'm not sure if the `except requests.exceptions.RequestException` (see CI logs) could be an issue? We usually only catch specific subclasses and let other exceptions bubble up.

It makes sense to intercept all failures as they're all, well, failures. Frankly at that point you might as well just `except Exception`.

The other endpoints are apparently more precise and do exclude some exception types but it seems mostly to provide different feedback for connection and http errors (why not) e.g. was it really a conscious decision to not handle `Timeout` errors and let those bubble up?

If you feel strongly about this it's easy enough to just catch the same two exceptions the other providers do, but I can't say I feel strongly about it.

**Comment by pga-odoo** [comment]
@AntoineVDV Official python SDK from mollie is also using `except requests.exceptions.RequestException` [see here](https://github.com/mollie/mollie-api-python/blob/master/mollie/api/client.py#L215)

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

usages of requests work fine, question is more in the details around it and that doesn't seem very security impacting.

**Comment by AntoineVDV** [comment]
@robodoo r+

File: addons/payment_mollie/models/payment_acquirer.py
Comment (odony) [inline addons/payment_mollie/models/payment_acquirer.py:64]: minor, but this is a complicated way of obtaining the `odoo.release.version` ;-)
```suggestion
        odoo_version = odoo.release.version
```
(after importing)

**Comment by Feyensv** [comment]
@robodoo rebase-ff

**Pull request closed without merging**
